src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -336,10 +313,14 @@ fn rustc(package: &Package, target: &Target, profile: &Profile,
     return rustcs.into_iter().map(|(mut rustc, kind)| {
         let name = package.name().to_string();
         let is_path_source = package.package_id().source_id().is_path();
-        let show_warnings = package.package_id() == cx.resolve.root() ||
-                            is_path_source;
-        if !show_warnings {
-            rustc.arg("-Awarnings");
+        let allow_warnings = package.package_id() == cx.resolve.root() ||
+                             is_path_source;
+        if !allow_warnings {
+            if cx.config.rustc_info().cap_lints {
+                rustc.arg("--cap-lints").arg("allow");
+            } else {
+                rustc.arg("-Awarnings");
```

Comment:
It's slightly amusing that the `--allow warnings` flag is being passed when `!allow_warnings`.